# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/webrender_bindings/RenderD3D11TextureHost.h
# Commit: 72d4bf5b093e
# Full Hash: 72d4bf5b093e9b00eed78bf4e51850d56df5ae6c
# Author: Matt Woodrow <mwoodrow@mozilla.com>
# Date: 2020-11-05 16:09:03
# Regressor Bug: 1673983
# File Overlap Count: 2
# Description:
#   Bug 1673983 - Add RenderTextureHostSWGL support to D3D11 RenderTextureHosts. r=lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D95799
# ==============================================================================

diff -r 99b2e5c55a25 -r 72d4bf5b093e gfx/webrender_bindings/RenderD3D11TextureHost.h
--- a/gfx/webrender_bindings/RenderD3D11TextureHost.h	Thu Nov 05 09:14:52 2020 +0000
+++ b/gfx/webrender_bindings/RenderD3D11TextureHost.h	Thu Nov 05 09:15:05 2020 +0000
@@ -8,7 +8,7 @@
 #define MOZILLA_GFX_RENDERD3D11TEXTUREHOST_H
 
 #include "GLTypes.h"
-#include "RenderTextureHost.h"
+#include "RenderTextureHostSWGL.h"
 
 struct ID3D11Texture2D;
 struct IDXGIKeyedMutex;
@@ -17,7 +17,7 @@
 
 namespace wr {
 
-class RenderDXGITextureHost final : public RenderTextureHost {
+class RenderDXGITextureHost final : public RenderTextureHostSWGL {
  public:
   explicit RenderDXGITextureHost(WindowsHandle aHandle,
                                  gfx::SurfaceFormat aFormat,
@@ -37,14 +37,32 @@
 
   RenderDXGITextureHost* AsRenderDXGITextureHost() override { return this; }
 
-  gfx::SurfaceFormat GetFormat() const { return mFormat; }
-
-  gfx::YUVColorSpace GetYUVColorSpace() const { return mYUVColorSpace; }
-
   gfx::ColorRange GetColorRange() const { return mColorRange; }
 
   ID3D11Texture2D* GetD3D11Texture2D();
 
+  // RenderTextureHostSWGL
+  gfx::SurfaceFormat GetFormat() const override { return mFormat; }
+  gfx::ColorDepth GetColorDepth() const override {
+    if (mFormat == gfx::SurfaceFormat::P010) {
+      return gfx::ColorDepth::COLOR_10;
+    }
+    if (mFormat == gfx::SurfaceFormat::P016) {
+      return gfx::ColorDepth::COLOR_16;
+    }
+    return gfx::ColorDepth::COLOR_8;
+  }
+  size_t GetPlaneCount() const override;
+  bool MapPlane(RenderCompositor* aCompositor, uint8_t aChannelIndex,
+                PlaneInfo& aPlaneInfo) override;
+  void UnmapPlanes() override;
+  gfx::YUVColorSpace GetYUVColorSpace() const override {
+    return mYUVColorSpace;
+  }
+
+  bool EnsureD3D11Texture2D(ID3D11Device* aDevice);
+  bool LockInternal();
+
  private:
   virtual ~RenderDXGITextureHost();
 
@@ -59,6 +77,11 @@
   RefPtr<ID3D11Texture2D> mTexture;
   RefPtr<IDXGIKeyedMutex> mKeyedMutex;
 
+  // Temporary state between MapPlane and UnmapPlanes.
+  RefPtr<ID3D11DeviceContext> mDeviceContext;
+  RefPtr<ID3D11Texture2D> mCpuTexture;
+  D3D11_MAPPED_SUBRESOURCE mMappedSubresource;
+
   EGLSurface mSurface;
   EGLStreamKHR mStream;
 
@@ -74,9 +97,11 @@
   bool mLocked;
 };
 
-class RenderDXGIYCbCrTextureHost final : public RenderTextureHost {
+class RenderDXGIYCbCrTextureHost final : public RenderTextureHostSWGL {
  public:
   explicit RenderDXGIYCbCrTextureHost(WindowsHandle (&aHandles)[3],
+                                      gfx::YUVColorSpace aYUVColorSpace,
+                                      gfx::ColorDepth aColorDepth,
                                       gfx::IntSize aSizeY,
                                       gfx::IntSize aSizeCbCr);
 
@@ -90,6 +115,22 @@
 
   bool SyncObjectNeeded() override { return true; }
 
+  // RenderTextureHostSWGL
+  gfx::SurfaceFormat GetFormat() const override {
+    return gfx::SurfaceFormat::YUV;
+  }
+  gfx::ColorDepth GetColorDepth() const override { return mColorDepth; }
+  size_t GetPlaneCount() const override { return 3; }
+  bool MapPlane(RenderCompositor* aCompositor, uint8_t aChannelIndex,
+                PlaneInfo& aPlaneInfo) override;
+  void UnmapPlanes() override;
+  gfx::YUVColorSpace GetYUVColorSpace() const override {
+    return mYUVColorSpace;
+  }
+
+  bool EnsureD3D11Texture2D(ID3D11Device* aDevice);
+  bool LockInternal();
+
  private:
   virtual ~RenderDXGIYCbCrTextureHost();
 
@@ -109,6 +150,12 @@
   // The gl handles for Y, Cb and Cr data.
   GLuint mTextureHandles[3];
 
+  // Temporary state between MapPlane and UnmapPlanes.
+  RefPtr<ID3D11DeviceContext> mDeviceContext;
+  RefPtr<ID3D11Texture2D> mCpuTexture[3];
+
+  gfx::YUVColorSpace mYUVColorSpace;
+  gfx::ColorDepth mColorDepth;
   gfx::IntSize mSizeY;
   gfx::IntSize mSizeCbCr;
 